serde-generate/runtime/python/bincode/__init__.py

```diff
@@ -0,0 +1,227 @@
+
+
+def encode_variant_index(value: int) -> bytes:
+    return int(value).to_bytes(4, "little", signed=False)
+
+
+def encode_str(value: str) -> bytes:
+    return encode_length(len(value)) + value.encode()
+
+
+def encode_bytes(value: bytes) -> bytes:
+    return encode_length(len(value)) + value
+
+
+def not_implemented():
+    raise NotImplementedError
+
+primitive_encode_map = {
+    st.bool: lambda x: int(x).to_bytes(1, "little", signed=False),
+
+    st.uint8: lambda x: int(x).to_bytes(1, "little", signed=False),
+    st.uint16: lambda x: int(x).to_bytes(2, "little", signed=False),
+    st.uint32: lambda x: int(x).to_bytes(4, "little", signed=False),
+    st.uint64: lambda x: int(x).to_bytes(8, "little", signed=False),
+    st.uint128: lambda x: int(x).to_bytes(16, "little", signed=False),
+
+    st.int8: lambda x: int(x).to_bytes(1, "little", signed=True),
+    st.int16: lambda x: int(x).to_bytes(2, "little", signed=True),
+    st.int32: lambda x: int(x).to_bytes(4, "little", signed=True),
+    st.int64: lambda x: int(x).to_bytes(8, "little", signed=True),
+    st.int128: lambda x: int(x).to_bytes(16, "little", signed=True),
+
+    st.float32: lambda x: not_implemented(),
+    st.float64: lambda x: not_implemented(),
+    st.unit: lambda x: b'',
+
+    st.char: lambda x: not_implemented(),
+    str: lambda x: encode_str(x),
+    bytes: lambda x: encode_bytes(x),
+}
+
+primitive_decode_map = {
+    st.bool: lambda content: (st.bool(int.from_bytes(content[:1], byteorder="little", signed=False)), content[1:]),
+
+    st.uint8: lambda content: (st.uint8(int.from_bytes(content[:1], byteorder="little", signed=False)), content[1:]),
+    st.uint16: lambda content: (st.uint16(int.from_bytes(content[:2], byteorder="little", signed=False)), content[2:]),
+    st.uint32: lambda content: (st.uint32(int.from_bytes(content[:4], byteorder="little", signed=False)), content[4:]),
+    st.uint64: lambda content: (st.uint64(int.from_bytes(content[:8], byteorder="little", signed=False)), content[8:]),
+    st.uint128: lambda content: (st.uint128(int.from_bytes(content[:16], byteorder="little", signed=False)), content[16:]),
+
+    st.int8: lambda content: (st.int8(int.from_bytes(content[:1], byteorder="little", signed=True)), content[1:]),
+    st.int16: lambda content: (st.int16(int.from_bytes(content[:2], byteorder="little", signed=True)), content[2:]),
+    st.int32: lambda content: (st.int32(int.from_bytes(content[:4], byteorder="little", signed=True)), content[4:]),
+    st.int64: lambda content: (st.int64(int.from_bytes(content[:8], byteorder="little", signed=True)), content[8:]),
+    st.int128: lambda content: (st.int128(int.from_bytes(content[:16], byteorder="little", signed=True)), content[16:]),
+
+    st.float32: lambda content: not_implemented(),
+    st.float64: lambda content: not_implemented(),
+
+    st.unit: lambda content: (None, content),
+    st.char: lambda content: not_implemented(),
```

Comment:
this could be just chr(content[0]), content[1:]

Reply:
No, unfortunately, `char` is supposed to be a unicode character -- taking between 1 and 4 bytes depending on the value.